### PR TITLE
Refactoring Loaders & Renders

### PR DESCRIPTION
**Pull request by WengerK**
When merged, should refactore also every other Twig Extension to use the new Drupal\bamboo_twig\TwigExtensionTwigExtensionBase.

**Loader**
run test with ```../../vendor/bin/phpunit --group bamboo_twig_loader --printer="\Drupal\Tests\Listeners\HtmlOutputPrinter"```

**Render**
run test with ```../../vendor/bin/phpunit --group bamboo_twig_render --printer="\Drupal\Tests\Listeners\HtmlOutputPrinter"```

 or more globally with ```../../vendor/bin/phpunit --group bamboo_twig --printer="\Drupal\Tests\Listeners\HtmlOutputPrinter"```

### CODE REVIEW

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  public function renderBlock($block_id, $params = []) {
+    $instance = $this->getPluginManagerBlock()->createInstance($block_id, $params);
+    return $instance->build($params);
+  }
+
+  /**
+   * Load a given block with or whitout parameters.
+   */
+  public function renderForm($module, $form, $params = []) {
+    return $this->getFormBuilder()->getForm('Drupal\\' . $module . '\Form\\' . $form, $params);
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php]: check if `\Form` should be `\\Form`.

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  /**
+   * Load a given block with or whitout parameters.
+   */
+  public function renderBlock($block_id, $params = []) {
+    $instance = $this->getPluginManagerBlock()->createInstance($block_id, $params);
+    return $instance->build($params);
+  }
+
+  /**
+   * Load a given block with or whitout parameters.
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php]: Update DocBlock

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+  /**
+   * Unique identifier for this Twig extension.
+   */
+  public function getName() {
+    return 'bamboo_twig_loader.twig.render';
+  }
+
+  /**
+   * Load a given block with or whitout parameters.
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php:38]: Update DocBlock (parameters)

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+   * @param string $region
+   *   The region to build.
+   * @param string $theme
+   *   (optional) The name of the theme to load the region. If it is not
+   *   provided then default theme will be used.
+   *
+   * @return array
+   *   A render array to display the region content.
+   */
+  public function renderRegion($region, $theme = NULL) {
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php:204]: I propose to remove this feature.
I don't see the usecase

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+   * Returns the render array for a single entity field.
+   *
+   * @param string $field_name
+   *   The field name.
+   * @param string $entity_type
+   *   The entity type.
+   * @param mixed $id
+   *   (optional) The ID of the entity to render.
+   * @param string $formatter
+   *   (optional) The formatter that should be used to render the field.
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php]: Add examples values

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+   * @param string $langcode
+   *   (optional) Language code to load translation.
+   *
+   * @return null|array
+   *   A render array for the field or NULL if the value does not exist.
+   */
+  public function renderField($field_name, $entity_type, $id = NULL, $formatter = NULL, $langcode = NULL) {
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php]: switch `$formatter` and `$langcode`

File: bamboo_twig_loader/src/TwigExtension/Render.php [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    }
+    else {
+      // We don't have the formatter view display and should fall back on
+      // default formatters.
+      $field_type_definition = $this->getFieldTypeManager()->getDefinition($entity->getFieldDefinition($field_name)->getType());
+      $display_options['type'] = $field_type_definition['default_formatter'];
+    }
+
+    if (isset($entity->{$field_name})) {
+      return $entity->{$field_name}->view($display_options);
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Render.php]: I prefer `$entity->get($field_name)`

File: bamboo_twig_loader/src/TwigExtension/Loader.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFunctions() {
+    return [
+      new \Twig_SimpleFunction('bamboo_load_entity', [$this, 'loadEntity']),
+      new \Twig_SimpleFunction('bamboo_load_field', [$this, 'loadField']),
+      new \Twig_SimpleFunction('bamboo_load_currentuser', [$this, 'loadCurrentUser']),
+      new \Twig_SimpleFunction('bamboo_load_image', [$this, 'loadImage']),
```
Comment (gido) [inline bamboo_twig_loader/src/TwigExtension/Loader.php:20]: add `bamboo_load_menu` ?

File: tests/bamboo_twig_test/templates/loader.html.twig [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  {% set filename = bamboo_load_field('filename', 'file', 1) %}
+  <div class="loader-field-file">{{ filename.value }}</div>
+  {% set username = bamboo_load_field('name', 'user', 1) %}
+  <div class="loader-field-user">{{ username.value }}</div>
+
+  <!-- Loader Image -->
+  <h2>Loader Image</h2>
+  <hr>
+  {% set image = bamboo_load_image('public://antistatique.png') %}
+  <div class="loader-image-uri">{{ image.getSource }}</div>
```
Comment (gido) [inline tests/bamboo_twig_test/templates/loader.html.twig]: replace by `image.getSource()` if this is a method

Comment (gido) [reply tests/bamboo_twig_test/templates/loader.html.twig]: or `image.source`

**Review by gido — CHANGES_REQUESTED** [review]

File: bamboo_twig_loader/src/TwigExtension/Render.php (hunk 3 above)
Comment (WengerK) [reply bamboo_twig_loader/src/TwigExtension/Render.php:38]: done

File: bamboo_twig_loader/src/TwigExtension/Loader.php (hunk 8 above)
Comment (WengerK) [reply bamboo_twig_loader/src/TwigExtension/Loader.php:20]: Create a issue for this here : https://github.com/antistatique/drupal-bamboo-twig/issues/7
People will chose.

**Pull request merged**